### DISCUSSION TITLE
Port the `Prior` class from `pymc-marketing` into the `pymc` repo

### DISCUSSION THREAD

**Discussion by drbenvincent** [pymc-devs/pymc3#7416] [Ideas]
I propose that we port the excellent `Prior` class by @wd60622 from the [`pymc-marketing`](https://github.com/pymc-labs/pymc-marketing) repo.

## What is it & what does it do?

The `Prior` class is basically a wrapper around pymc distributions. It has a number of useful properties:
* In packages built on top of `pymc`, it is a way of allowing users to define custom priors in pre-build models, or model factory functions/classes. This supersedes a previous way of doing it based on parsing dicts. This is exactly what it's doing in `pymc-marketing`. I would definitely bring it into [`CausalPy`](https://github.com/pymc-labs/CausalPy). And in theory it could also be used in packages like `bambi` - though I have no idea if there's a desire to do that.
* It also acts as a very nice abstraction layer / API for a user to specify possibly complex prior knowledge, such as hierarchies. Users can flip between centered and non-centred parameterisations simply by changing a kwarg.

Bringing it in to `pymc` would allow _an additional_ way to specify priors. This would be most appealing in hierarchical modeling situations, for example.

## More info
* The `Prior` api docs in pymc-marketing: https://www.pymc-marketing.io/en/latest/api/generated/pymc_marketing.prior.html
* Example notebook demonstrating it's use: https://www.pymc-marketing.io/en/latest/notebooks/general/model_configuration.html

👆Please do check out these pages to get properly up to speed

## Options
I see at least 4 possible options, and it would be great to get some input from folks.

1. Do nothing, leave it in `pymc-marketing`. This would then remain out of reach of other packages built on top of pymc.
2. Move it into it's own (small) repo which could be used as a dependency that other packages built on top of pymc which want to enable users to specify custom priors.
3. Move it into [`pymc-experimental`](https://github.com/pymc-devs/pymc-experimental) with potential to move it into `pymc` in the future
4. Move it into `pymc` now.

(Personally, I like options 4, 2, and 3, in that order)

**Comment by ricardoV94** [comment]
We already have a way to define variables outside of a model, what if we try to use that?

```python
x = pm.Normal.dist(0, 1)

with pm.Model() as m:
  m.register_rv(x, name="x")
```

Pros: allows as much expressivity as PyMC models and uses almost same API.

Cons: 
1. Allows as much expressivity as PyMC models
1. Need to get dims out of the model (that's the only thing you can't do now outside), but I don't think that's too terrible. The helper is doing it as well
2. I think PyTensor graphs are serializable, but would need to check. Not as immediate as a text string is on that front

Alternative what about lambda functions / callables / functools.partial?

Also may be worth checking the approach @jessegrabowski took with Statspace module, which is kind of "user defines a model that fills in the blanks, with defaults for things that are not specified".

**Comment by juanitorduz** [comment]
I like the `Prior` concept and design, but I am heavily biased as I contributed to reviewing it 🙈 . Still, I think it would be a good addition (at least) for `pymc-experimental`.

**Reply by ricardoV94** [reply]
What do you like about it vs PyMC vanilla API?

**Reply by juanitorduz** [reply]
Again, totally subjective, but I really like how we can build like “components” and handle dimensions as described in https://www.pymc-marketing.io/en/stable/notebooks/mmm/mmm_components.html 

We can do this in pure PyMC, but it feels a bit user-friendlier  (but again, I'm used to it so I am underestimating the work of adding this to the pymc learning path)

PyMC marketing is a great use case because we want to provide simple building blocks. Now that I am thinking about it maybe PyMC is not at the same level of abstraction / flexibility... I don't know now 😄

**Reply by ricardoV94** [reply]
Prior sounds a lot like a partial?

```python
{'alpha': Prior("Gamma", mu=2, sigma=1), 
'lam': Prior("HalfNormal", sigma=1)}
````


```python
import functools.partial as Prior

{'alpha': Prior(pm.Gamma, mu=2, sigma=1),
 'lam': Prior(pm.HalfNormal, sigma=1)}
````

This doesn't require strings. But isn't as easily serializable as text which I think was one of the needs in pymc-marketing?

Or we don't want users to have to use PyMC at all? But that makes it harder to find the arguments.

**Reply by williambdean** [reply]
The Prior class's main benefit is the explicit dims outside of a model context which then supports the broadcasting. PyMC requires a model context in order to specify dims to my knowledge.
`Prior("Normal", mu=Prior("Normal", dims="channel"), sigma=Prior("HalfNormal", dims="channel"), dims=("channel", "geo"))`

Also think additional benefits of class methods for various applications which won't come with partial. Those are: 
- preliz
- to_graph
- constrain
- sample_prior

Though we do have check on the strings with the available PyMC distributions, maybe use of distributions at initialization would be more explicit. i.e `Prior(pm.Gamma, mu=2, sigma=1, dims="channel")`

**Reply by ricardoV94** [reply]
Yeah those go beyond partial. Thanks.

But any reason you can't do those with the `.dist` objects or wouldn't want to?  sample_prior is just pm.draw? plotting and preliz must be compatible with . dist objects. Plotting was rejected in another discussion because "arviz" or "preliz" does it or something. But if we're adding the functionality anyway that's moot.

The only one that's trickier is dims. We could allow dims outside since pymc supports symbolic shapes. We would just set them to the correct value when registering them in the model

**Reply by jessegrabowski** [reply]
If we add named dims to pytensor it would solve all of our problems :)

**Reply by williambdean** [reply]
Yes, seems like dims in pytensor could power this functionality as well. Is there any scoping of that? What would be a realistic timeline to implement that?

**Reply by ricardoV94** [reply]
long

**Comment by ricardoV94** [comment]
Just a point on this:

> Bringing it in to pymc would allow an additional way to specify priors. This would be most appealing in hierarchical modeling situations, for example.

An additional way is not necessarily a plus. Users have one more thing to learn / try to understand when they come across examples. This assumes an overlap between pymc users and those other libraries

It's also limited in different ways that may lead us to more complexity or reinvent pymc v2 with strings: e.g 
  * Do we want an option to share parameters across priors? If so, how does that look like?
  * Is it fine to exclude distribution factories like Censored/Truncated/CustomDist? If not, how will the API be made to work for those?
  * Do we want to forbid distributions from pymc-experimental/ third party libraries? If not, how do we support it?

Re: the centered/non-centered is a specific thing that may make more sense as a `pm.NonCentered` factory or the specialized `NonCenteredNormal` for the 99% use cases? I don't think it's an integral part of this proposal?

**Comment by ricardoV94** [comment]
Also curious what @tomicapretto is doing in over bambi, and what are the pain points

**Reply by tomicapretto** [reply]
We have had an abstraction for priors in Bambi since a long time ago, but it's much simpler than what Will implemented in pymc-marketing. In our case, it's just a structure that stores some data that we use to build PyMC distributions when we are writing the PyMC model. From what I can see, the Prior class in pymc-marketing does _a lot_ more.

I'll try to summarise my thoughts here.

What I like

* It makes it easier to work with hierarchies both in a PyMC model and outside of a PyMC model.
* It seems it handles dims pretty well.
* All the convenient utilities that are already implemented (those that Will mentioned)

What I don't like

* `Prior` is not the best name. It's actually a `Distribution`. Whether it's a prior or not depends on the context. Let's say this is a nitpick.
* The `transform` argument. I would like to do something like `sigmoid(Distribution(...))`.
* The `centered` argument. I know it's very convenient and it's used a lot. But:
  * For which distributions is it implemented? All the location-scale distributions or just Normal? 
  * If implemented for other distributions, how is it implemented? Does it make sense? For example, how does it work with `Prior("Beta")`.
  * Again, I would like it to be something like `NonCentered(Distribution(...))`. To understand where it comes from, I have a similar feeling towards `ZeroSumNormal`. To me it should be `ZeroSum(MvNormal(...))` or `ZeroSum(Normal(shape = n))` where n > 1.
* Other syntax for users to learn. I know this is _additive_, but I can imagine users will be confused by seeing different syntaxes being used in different examples. 


In my ideal world, users should be able to do this with PyMC or PyTensor variables _outside_ a PyMC model. 

```python

mu = pm.Normal()
sigma = pm.InvGamma()
beta = pm.Normal(mu, sigma, dims="geo")
beta_star = pm.NonCentered(beta)

beta_star.to_graphviz()
```

---

With all that said, I _do_ think the Prior class is a great addition to PyMC Marketing, that it makes the devs life easier and it can help users to build a PyMC model using components from PyMC Marketing. I'm just hesitant to include `Prior` (as it is) in PyMC.

**Reply by williambdean** [reply]
Yeah, I agree with these cons.
It would be ideal that transformations work on the classes directly. However, from historical context, this is just a first step away from the dictionary class that was used for prior distributions.
I've opened this pytensor issue to get the ball rolling there: https://github.com/pymc-devs/pytensor/issues/954

**Comment by ricardoV94** [comment]
Is this something like prior?

```python
class Prior:
  def __init__(self, dist, name, *args):
    ... # save

  @staticmethod
  def _eval(arg):
    if isinstance(arg, Prior):
      return arg.dist(
        arg.name, 
        *(_eval(arg) for arg in args), 
        **dict(kwarg=_eval(arg) for kwarg, arg in self.kwargs)
      )
    else:
      return arg

  def eval(self, coords=None, model=UNSET):
    # Unset makes nested models
    # None makes standalone models
    with pm.Model(coords=coords, model=model) as m:
      return _eval(self)

  def prior(self, draws=1000):
    return pm.draw(self.eval(model=None))

  def plot(self, draws=1000):     
    return pm.plots.plot_dist(
      self.prior(draws=draws)
    )
```

**Reply by williambdean** [reply]
Similar. Names are autogenerated based on the child parameters, shapes are handled based on the explicit dims, and post creation transformation is supported with a `transform` parameter, and non-centered / centered parameterization is supported.

The eval / create_variable (for non-centered) is only a few lines
https://github.com/pymc-labs/pymc-marketing/blob/main/pymc_marketing/prior.py#L411-L423

One major difference here is that name is not part of initialization. Only creation. The Prior class is only the instructions to create a variable.

Everything is in this single file for reference: https://github.com/pymc-labs/pymc-marketing/blob/main/pymc_marketing/prior.py

**Reply by ricardoV94** [reply]
It's easy to define the name later. My pseudo-code handles dims already, if they are one of the kwargs to Prior.

It still seems limited in that you can't use stuff like pm.Censored or reuse parameters across variables. Maybe the later is not a real need?

It at least avoids the restriction that it must be a distribution in PyMC

**Reply by williambdean** [reply]
Doesnt seem to handle case where parameter needs to be put through dimshuffle

Dont think it is too hard to support general PyMC distributions. The actual distribution is just lookup upon init.

No support for Censored as it wasnt a case supported for pymc marketing. 
One generalization I had in mind was to support parameters that have create_variable and dims property. This would help support r2d2, hsgp, and various custom random variables

**Reply by ricardoV94** [reply]
What do you mean a parameter has to be put into a DimShuffle? When/why is that needed? 

You're trying to align dims? That's interesting... 🤔

**Reply by ricardoV94** [reply]
> One generalization I had in mind was to support parameters that have create_variable and dims property. This would help support r2d2, hsgp, and various custom random variables

What do you have in mind that would look like?

**Reply by williambdean** [reply]
Child variables will be reshaped upon creation in order to work with parent dims. For instance,
```python
dist = Prior(
    "Normal", 
    mu=Prior("Normal", dims="dim2"), 
    dims=("dim1", "dim2", "dim3"),
)
samples = dist.sample_prior(coords={"dim1": [1, 2, 3], "dim2": ["A", "B"], "dim3": ["X", "Y", "Z"])
```

The user doesn't have to think about these transformations. This makes changing of mu shape just changing the dims of Prior.
This happens here: https://github.com/pymc-labs/pymc-marketing/blob/717702a1cc20dbad55d97b9020c3ecd086d74366/pymc_marketing/prior.py#L411-L416

> What do you have in mind that would look like?

Have the HSGP logic built out here: https://github.com/pymc-labs/pymc-marketing/pull/829 
example API in PR description. Though it isn't a Prior class, it has `dims` and `create_variable` method which acts the same. 

r2d2 is less scoped but might look like this: https://github.com/pymc-labs/pymc-marketing/issues/381#issuecomment-2208246893

**Reply by ricardoV94** [reply]
The dim align thing is something we really want to in PyTensor long term. It's not just Dimshuffle but also things like `a.dot(b, dims=("city"))`, regardless of whether they are.

Maybe you would like to think about that once you're done here ;)

Btw does Prior handle cases where dims in the prior get "consumed"? Like Categorical has a p vector that doesn't translate to any dims in the Categorical itself.

**Reply by williambdean** [reply]
> The dim align thing is something we really want to in PyTensor long term. It's not just Dimshuffle but also things like `a.dot(b, dims=("city"))`, regardless of whether they are.
> 
> Maybe you would like to think about that once you're done here ;)

Would be pretty cool functionality

> Btw does Prior handle cases where dims in the prior get "consumed"? Like Categorical has a p vector that doesn't translate to any dims in the Categorical itself.

I am not sure. Can you write a pymc example?

**Reply by ricardoV94** [reply]
> I am not sure. Can you write a pymc example?

```python
with pm.Model() as m:
  p = pm.Dirichlet("p", [1, 1, 1], dims=["probs"])
  # p = pm.Dirichlet("p", [1, 1, 1], dims=["trial", "probs"])  # Another option
  y = pm.Categorical("y", p=p, dims=["trial"])
```

**Reply by williambdean** [reply]
> > I am not sure. Can you write a pymc example?
> 
> 
> 
> ```python
> 
> with pm.Model() as m:
> 
>   p = pm.Dirichlet("p", [1, 1, 1], dims=["probs"])
> 
>   # p = pm.Dirichlet("p", [1, 1, 1], dims=["trial", "probs"])  # Another option
> 
>   y = pm.Categorical("y", p=p, dims=["trial"])
> 
> ```

Would have to be explicit with trial. Ie commented option
Though this might cause error with current iteration as it would be transposed. 
Is there something that signals this type of distribution in Categorical class? What other distributions are like that?

**Reply by ricardoV94** [reply]
Yes distribution op has a numpy-like signature `(p)->()` meaning the core case has a vector->scalar form. Multinomial has signature: `(),(p)->(p)`: https://www.pymc.io/projects/docs/en/stable/learn/core_notebooks/dimensionality.html#multivariate-distribution-example

**Reply by williambdean** [reply]
> Yes distribution op has a numpy-like signature `(p)->()` meaning the core case has a vector->scalar form. Multinomial has signature: `(),(p)->(p)`: https://www.pymc.io/projects/docs/en/stable/learn/core_notebooks/dimensionality.html#multivariate-distribution-example

Thats good. Is it exactly numpy standard? Are there parsing utilities? Does those exist in either numpy or pytensor

**Reply by ricardoV94** [reply]
This is using numpy vectorize signature, but it's a PyTensor thing. You should have it as a property of `pm.Normal.rv_type.signature` or something

**Reply by williambdean** [reply]
Nice. I was just exploring it a bit and created an issue here: https://github.com/pymc-labs/pymc-marketing/issues/841

Explicit dims work but could probably be relaxed

Was noticing this behavior of Categorical. Is this weird / a bug?
```python
p = pm.Dirichlet.dist(a=[1, 1, 1])
x = pm.Categorical.dist(p=p, shape=(5, 3)) # eval works
y = pm.Categorical.dist(p=p[:, None], shape=(5, 3)) # eval also works but unintuitive 
z = pm.Categorical.dist(p=p[:, None], shape=(3, 5))  # eval doesn't work because of TypeError shape mismatch
```
Would think padding p would make variable of shape=(3, 5) rather than (5, 3) but that causes shape mismatch

**Reply by ricardoV94** [reply]
It behaves as expected even if it may be a bit surprising. Notice the shape of the last dimension of `p` can be anything. This is valid:

```python
p = pm.Dirichlet.dist(a=np.ones(100))
x = pm.Categorical.dist(p=p, shape=(5, 3)) # (5, 3) categorical draws with values between 0-99
```

Your x is totally fine, it's a (5,3) categorical draws with values between 0-2
Your y is wishy wash. Technically the only valid values is if p was 1 for all entries, which would give you (5, 3) deterministic categorical draws of 0. This would be valid:

```python
p = pm.Dirichlet.dist(a=[1], shape=(3, 1))
y = pm.Categorical.dist(p=p, shape=(5, 3))
assert y.eval().max() == 0
```
Your case is technically wrong at runtime, because the probabilities are not always 1 (which in this case also means, they don't add up to 1). In this case Categorical could raise an Error, but we don't. The behavior is undefined, and in this case can return (0, 1), even though 1 should not be possible.

Your z variable is wrong, because of numpy broadcsating. You have an input with shape (3,1), the 1 (the last dimension) is a core dimension, so it doesn't matter for broadcasting. So you have a batch of (3,) p and are trying to use that for a variable with shape (3, 5), which doesn't match. It's similar to if you tried to run the following multinomial:

```python
pm.Multinomial.dist(n=np.ones((3,), p=np.ones(100)/100, shape=(3, 5,))
```
Or 
```python
pm.Multinomial.dist(n=np.ones((3,), p=np.full(shape=(3, 5), fill_value=1/5))
```

`shape` must also broadcast left to right, just like parameters. This is exactly the same as in numpy. That notebook I linked to goes over this in more detail: https://www.pymc.io/projects/docs/en/stable/learn/core_notebooks/dimensionality.html

**Reply by williambdean** [reply]
Ah. Yeah, makes sense. See my mistake. 
Thanks for the explanation

**Comment by jessegrabowski** [comment]
To add very little to this covo, `statespace` hasn't implemented the default priors yet, because I couldn't decide on an API for handling "model free" priors. The `.dist` with `model.register_rv` is close, but the pain points were 1) dims, and 2) "compound" priors, in my case diagonal covariance matrices. `Prior` handles both of these naturally, which is nice.

**Reply by williambdean** [reply]
Do you have some links for where this takes place for reference @jessegrabowski

**Reply by jessegrabowski** [reply]
I had the idea that models should implement a [default_priors](https://github.com/pymc-devs/pymc-experimental/blob/87d4aea8d3c3003e641e0de6d5fcb861a224c189/pymc_experimental/statespace/core/statespace.py#L409) property that would give a dictionary of var_name:prior pairs, which could be used by a (not yet implemented) method `add_default_priors`. That could be used in `_insert_random_variables` [here ](https://github.com/pymc-devs/pymc-experimental/blob/87d4aea8d3c3003e641e0de6d5fcb861a224c189/pymc_experimental/statespace/core/statespace.py#L641) instead of raising an error if a required RV is missing.

**Reply by ricardoV94** [reply]
@jessegrabowski your approach sounds a lot like pattern substitution.

Indulge my use of S-expressions to represent nested callables transparently 

Model = (pm.Normal, "y", "mean", (pm.HalfNormal, "sigma", "std"))

Then user needs to provide "mean" and "std" in this example or you could give defaults.

model_config = {
  "mean": (pm.Normal, 0, 1),
  "std": 10,
}

To build the model you start evaling the S-expressions, replacing place_holders by expressions in model_config.

Prior does something like S expression (and can be used as an alternative), by allowing you to define a lazy distribution to plug-in in the model.

I guess in STS you're doing it with clone_replace on some root/dummy variables?

**Comment by ricardoV94** [comment]
Another thing worth discussing. Do we want this to allow defining intermediate graphs. Say library offers model with components A -> B -> C and implements A and C and user can customize B. 

I don't think any of the existing suggestions would work for this use case, but can perhaps be tweaked for it.

Stuff like prior and plotting wouldn't be possible for B objects obviously

**Reply by williambdean** [reply]
I don't really follow this. Do you have some pseudo-code or example in mind?

**Reply by ricardoV94** [reply]
In a model like 

```python
A = pm.Normal("A")
B = pm.Gamma("B", mu=A, dims="country")
C = pm.Norma("C", mu=B[idx], dims="trial", observed=data)
```

Would it be useful for the user to control what B looks like?

(I'm not sure)

**Comment by ricardoV94** [comment]
Btw I think we definitely want to figure out something like this. Since there are 3 distinct libraries that would depend on this and there's probably some room for tweaking I would suggest adding it first to experimental and once all 3 libraries are happy with it and haven't felt the need to tweak it further we move to pymc?

**Reply by williambdean** [reply]
Think this is fair solution. Definitely would like to build out the solution a bit more as it came pretty naturally from the dictionary configurations of pymc-marketing which didn't have much functionality to begin with. 

We just need to bring in pymc-experimental with pymc-marketing again. Any opposition to this @juanitorduz ? The dependencies seems to just be pymc and sklearn so there shouldn't be any conflicts or restrictions

**Comment by ferrine** [comment]
I had something like this in the past. What I came up with is creating a closure that is provided by user and executed inside the model context

```python
from functools import partial

def priors_fn(**config) -> dict[str, pt.TensorVariable]:
   ...

class PriorsCls(pydantic.BaseModel):
    ...
    def __call__(self) -> dict[str, pt.TensorVariable]:
        ...

def lazy_get(cfg: dict[K_t, V_t], key: str, on_empty: Callable[[], V_t], on_value: Callable[[V_t], V_t]=lambda x: x) -> V_t:
    value = cfg.get(key)
    if key is None:
        value = on_empty()
    else:
        value = on_value(value)
    return value

def lazy_get_deterministic(cfg: dict[str, pt.TensorValue], key: str, on_empty: Callable[[], pt.TensorValue], possible_dims: list[str]) -> pt.TensorValue:
    return lazy_get(cfg, key, on_empty, lambda v: pm.Deterministic(key, v, dims=infer_dims(v, possible_dims=possible_dims)))

def model_builder(priors_fn):
    with pm.Model(coords=...) as model:
        with pm.Model("user_priors"):
            priors = priors_fn()
            cfg = partial(lazy_get, priors)
            cfg_deterministic = partial(lazy_get_deterministic, priors)
        b = cfg("b", 
            lambda: pm.Normal("b", dims="c"), 
            lambda v: pm.Deterministic("b", v, dims=infer_dims(v, possible_dims=["a", "b", "c"])) # infer_dims might be needed if dim number might be overriden
        )
        c = cfg_deterministic("g", lambda: pm.Normal("g", dims="c"), possible_dims=["a", "b", "c"])
```

Effectively, the user facing protocol is the parametrized lambda function, with the only assumption to be called under a model context.
On the model side, the requirement is 
- to handle missing keys properly 
- verify the user follows dims protocol when passing uncertain tensor to the model. That is, runtime check for dims match with each other and wrapping the output in deterministic for consistent InferenceData analysis.

I believe some nice api over
```python
cfg("b", 
   lambda: pm.Normal("b", dims="c"), 
   lambda v: pm.Deterministic("b", v, dims=infer_dims(v, possible_dims=["a", "b", "c"]))
)
```
can be also proposed, but the nature still seems single use inside the model and that is not a requirement to make use of it. What about

```python
def infer_dims(value, possible_dims: list[str]):
    if value.ndim:
        # slice from the end
        return possible_dims[-value.ndim:]
    else:
        # we have scalar
        return None
```

**Reply by williambdean** [reply]
So it seems like the ability of having distributions outside of a model would be pretty useful then. 

The `Prior` class just executes the `create_variable` method from within the `pm.Model` context so this similar to storing a bunch of callables. Don't find storing a bunch of lambda functions as the most user friendly solution though
